### ISSUE TITLE
[HOLD for payment 2024-11-28] [$250] Expense - Payment sound triggered prematurely when submitting a held expense.

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#50572]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.46-5
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** biruknew45+1482@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
1. Go to https://staging.new.expensify.com/
2. Create a workspace
3. Navigate to the workspace chat
4. Submit an expense
5. Right-click on the submitted expense and select hold
6. Click the green "Pay" button
7. Observe that the payment sound is triggered and an additional "Confirm payment amount" modal appears

## Expected Result:
For this case the sound indicating a payment should be triggered after the payment is successfully completed. If the expense is on hold, the sound should be delayed until the confirmation is made and finalized

## Actual Result:
The payment sound is triggered when the green pay button is clicked, even though the payment is not yet finalized

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

<details>
  <summary>Add any screenshot/video evidence</summary>

  
</details> 

https://github.com/user-attachments/assets/75f672e8-1b8f-4f9e-ba76-68dc6bd3f297

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021846013804908725875</li>
        <li>Upwork Job ID: 1846013804908725875</li>
        <li>Last Price Increase: 2024-10-22</li>
<li>Automatic offers: </li>
<ul>
<li>mkzie2 | Contributor | 104560763</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @abdulrahuman5196</details>

**Comment by mkzie2** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The payment sound is triggered when the green pay button is clicked, even though the payment is not yet finalized

### What is the root cause of that problem?

We're play sound when after selecting payment type in `SettlementButton` regardless of the issue is on hold or not:

https://github.com/Expensify/App/blob/420e780c359d861f7a43fda3441104c42ac39225/src/components/SettlementButton/index.tsx#L210

### What changes do you think we should make in order to solve the problem?


We should not play sound in `SettlementButton` when the expense is on hold but play it after confirming in the hold modal instead.

1. Introduce `shouldPlaySound` in `SettlementButton` to determine whether to play sound or not
2. In `ReportPreview` and `MoneyReportHeader`, pass `shouldPlaySound={isAnyTransactionOnHold}`, or `hasHeldExpenses`:

https://github.com/Expensify/App/blob/5b385166612261772ef26e628c5ad60632914454/src/components/MoneyReportHeader.tsx#L146

https://github.com/Expensify/App/blob/5b385166612261772ef26e628c5ad60632914454/src/components/ReportActionItem/ReportPreview.tsx#L204-L205

3. Play the sound when confirming the hold modal:

https://github.com/Expensify/App/blob/c0d26c570c488d7f6bc289996e1c60e30d3bf97f/src/components/ProcessMoneyReportHoldMenu.tsx#L66

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Pressing the pay button plays the sound even though a confirm modal shows.

### What is the root cause of that problem?
We always play the sound without checking whether there is a held expenses or not.
https://github.com/Expensify/App/blob/517ffaaf4aad021872c8c9f1b7e73ae8366fb5ca/src/components/SettlementButton/index.tsx#L210-L211

### What changes do you think we should make in order to solve the problem?

Instead of playing the sound when pressing the SettlementButton, we can play it inside the pay function, such as payMoneyRequest and payInvoice. We will follow the same pattern as completing a task.
https://github.com/Expensify/App/blob/517ffaaf4aad021872c8c9f1b7e73ae8366fb5ca/src/libs/actions/Task.ts#L408-L409

**Comment by RachCHopkins** [comment]
Easy to repro. It sounds when you hit the green button, not when you confirm.

**Comment by RachCHopkins** [comment]
@abdulrahuman5196 we have a number of early proposals here - do any look good to you?

**Comment by abdulrahuman5196** [comment]
checking now

**Comment by abdulrahuman5196** [comment]
@mkzie2 's proposal here https://github.com/Expensify/App/issues/50572#issuecomment-2405181970 looks good and works well. One suggestion for the PR would be to try and see if we can avoid introducing a new param in the settlement button which is already bloated, instead we can check if the expense is on hold inside the settlement button itself. I can see we already have chatReport there which we can re-use.
All proposals here are almost the same, so going with the first proposal.

🎀 👀 🎀
C+ Reviewed

**Comment by bernhardoj** [comment]
> All proposals here are almost the same, so going with the first proposal.

I don't think my proposal is almost the same as the others. My proposal suggesting to follow the same pattern as creating task, which doesn't require any conditional logic.

**Comment by carlosmiceli** [comment]
@abdulrahuman5196 After looking at @bernhardoj's proposal, I agree that it's not the same. What's the case for the one you selected?

@mkzie2 please let's hold a bit on the PR 🙏

**Comment by bernhardoj** [comment]
With my solution, first we will remove the `playSound` from SettlementButton.
https://github.com/Expensify/App/blob/66112d13bd47ed3d9b029e77b8d494e13808a594/src/components/SettlementButton/index.tsx#L211-L212

Then, play the sound inside `IOU.payMoneyRequest` (or payInvoice, sendMoneyElsewhere, and sendMoneyWithWallet).
https://github.com/Expensify/App/blob/66112d13bd47ed3d9b029e77b8d494e13808a594/src/libs/actions/IOU.ts#L7658-L7675

So, we don't play the sound when the button is pressed, but when the request is pay made.

**Comment by abdulrahuman5196** [comment]
I agree its not the same as the accepted proposal. I think I had initially misunderstood the proposal.
Anyways I will give a check on this again and update with the final result today.

**Comment by abdulrahuman5196** [comment]
I have checked @bernhardoj  proposal. And IMO, the concern I have is, removing `playSound(SOUNDS.SUCCESS); ` from `SettlementButton` and adding it to places where we would make the API call would cause us to add the playSound in multiple places in IOU functions which would be a duplicate code. And I am not a fan of having UX interaction like playSound to happen from API calling place which would break the UX and network separation. 

So I still think the accepted proposal is better. @carlosmiceli. But anyways do let us know your thoughts as well.

**Comment by carlosmiceli** [comment]
Understood, @abdulrahuman5196, I like your logic here, I just wanted to confirm that proposals were distinct and a clear reasoning behind the selected one. @mkzie2 feel free to proceed and thanks for understanding with the delay.

**Comment by RachCHopkins** [comment]
@abdulrahuman5196 I'm just waiting on your checklist here before I issue payment.

**Comment by carlosmiceli** [comment]
Not overdue Melv

**Comment by abdulrahuman5196** [comment]
# BugZero Checklist:

- [ ] **[Contributor]** Classify the bug:

<details>
<summary>Bug classification</summary>


Source of bug:
  - [ ] 1a. Result of the original design (eg. a case wasn't considered)
  - [ ] 1b. Mistake during implementation
  - [ ] 1c. Backend bug
  - [ ] 1z. Other:

Where bug was reported:
  - [ ] 2a. Reported on production
  - [ ] 2b. Reported on staging (deploy blocker)
  - [ ] 2c. Reported on both staging and production
  - [ ] 2d. Reported on a PR
  - [ ] 2z. Other:

Who reported the bug:
  - [ ] 3a. Expensify user
  - [ ] 3b. Expensify employee
  - [ ] 3c. Contributor
  - [ ] 3d. QA
  - [ ] 3z. Other:

</details>

- [ ] **[Contributor]** The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake.

    Link to comment:

- [ ] **[Contributor]** If the regression was CRITICAL (e.g. interrupts a core flow) A discussion in [#expensify-open-source](https://app.slack.com/client/E047TPA624F/C01GTK53T8Q) has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner.

    Link to discussion:

- [ ] **[Contributor]** If it was decided to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps using the template below to ensure the same bug will not reach production again.

<details>
<summary>Regression Test Proposal Template</summary>


- [ ] **[BugZero Assignee]** Create a GH issue for creating/updating the regression test once above steps have been agreed upon.

    Link to issue:

## Regression Test Proposal
### Precondition:


-

### Test:


1.

Do we agree 👍 or 👎


</details>

**Comment by abdulrahuman5196** [comment]
@RachCHopkins Added the checklist above.

**Comment by RachCHopkins** [comment]
Payment Summary:

- Contributor: @mkzie2 to be paid $250 via Upwork
- Contributor+: @abdulrahuman5196 to be paid $250 via NewDot Manual Request

Upwork job [here](https://www.upwork.com/jobs/~021846013804908725875)

**Comment by RachCHopkins** [comment]
Contributor has been paid, the contract has been completed, and the Upwork post has been closed.

**Comment by JmillsExpensify** [comment]
$250 approved for @abdulrahuman5196
